diff --git a/build/mongodl.py b/build/mongodl.py
index 9baff98d3..07187f5f6 100644
--- a/build/mongodl.py
+++ b/build/mongodl.py
@@ -33,6 +33,7 @@ DISTRO_ID_MAP = {
     'opensuse': 'sles',
     'redhat': 'rhel',
     'rocky': 'rhel',
+    'almalinux': 'rhel',
 }
 
 DISTRO_VERSION_MAP = {
@@ -61,6 +62,8 @@ DISTRO_ID_TO_TARGET = {
         '7.*': 'rhel70',
         '8': 'rhel80',
         '8.*': 'rhel80',
+        '9': 'rhel80',
+        '9.*': 'rhel80',
     },
     'sles': {
         '10.*': 'suse10',
